qiskit_experiments/library/characterization/analysis/zz_ramsey_analysis.py

```diff
@@ -0,0 +1,253 @@
+            curve.guess.frequency(data_1.x, data_1.y),
+        ]
+        base_guesses = [
+            curve.guess.constant_sinusoidal_offset(data_0.y),
+            curve.guess.constant_sinusoidal_offset(data_1.y),
+        ]
+
+        def rough_sinusoidal_decay_constant(
+            x_data: np.ndarray, y_data: np.ndarray, bounds: Tuple[float, float]
+        ) -> float:
+            """Estimate the decay constant of y_data vs x_data
+
+            This function assumes the data is roughly evenly spaced and that
+            the y_data goes through a few periods so that the peak to peak
+            value early in the data can be compared to the peak to peak later
+            in the data to estimate the decay constant.
+
+            Args:
+                x_data: x-axis data
+                y_data: y-axis data
+                bounds: minimum and maximum allowed decay constant
+
+            Returns:
+                The bounded guess of the decay constant
+            """
+            x_median = np.median(x_data)
+            i_left = x_data < x_median
+            i_right = x_data > x_median
+
+            y_left = np.ptp(y_data[i_left])
+            y_right = np.ptp(y_data[i_right])
+            x_left = np.average(x_data[i_left])
+            x_right = np.average(x_data[i_right])
+
+            # Now solve y_left = exp(-x_left / tau) and
+            # y_right = exp(-x_right / tau) for tau
+            denom = np.log(y_right / y_left)
+            if denom < 0:
+                tau = (x_left - x_right) / denom
+            else:
+                # If amplitude is constant or growing from left to right, bound
+                # to the maximum allowed tau
+                tau = bounds[1]
+
+            return max(min(tau, bounds[1]), bounds[0])
+
+        user_opt.p0.set_if_empty(
+            tau=rough_sinusoidal_decay_constant(curve_data.x, curve_data.y, user_opt.bounds["tau"]),
+            amp=y_ptp / 2,
+            phase=0.0,
+            freq=float(np.average(freq_guesses)),
+            base=np.average(base_guesses),
+            zz=freq_guesses[1] - freq_guesses[0],
+        )
+
+        return user_opt
+
+    def _evaluate_quality(self, fit_data: CurveFitResult) -> Union[str, None]:
+        """Algorithmic criteria for whether the fit is good or bad.
+
+        A good fit has:
+            - a reduced chi-squared lower than three
+            - an error on the frequency smaller than the frequency
+
+        Args:
+            fit_data: The fit result of the analysis
+
+        Returns:
+            The automated fit quality assessment as a string
+        """
+        freq = fit_data.ufloat_params["freq"]
+        zz = fit_data.ufloat_params["zz"]
+        amp = fit_data.ufloat_params["amp"]
+        base = fit_data.ufloat_params["base"]
+
+        rough_freq_magnitude = 1 / (fit_data.x_range[1] - fit_data.x_range[0])
+
+        criteria = [
+            amp.nominal_value > 5 * amp.std_dev,
```

Comment:
Can we use this helper function? This has some edge case handling.
https://github.com/Qiskit/qiskit-experiments/blob/219af364f10486bfa6ed2bd57b1a1b1552c9fbae/qiskit_experiments/curve_analysis/utils.py#L33-L37

Reply:
Sure, see da8aeb5b5def5743dc012eeed5a56eeab2a2c17f. I used conditionals because I thought that was more readable -- here I was comparing the error of parameters to the nominal values of different parameters rather than the more standard case of compare error and value of a single parameter -- but we can use the helper to address the edge cases by making use of the `absolute` keyword.